src/vhost.erl

```diff
@@ -0,0 +1,167 @@
+  get_limits/1,
+  get_metadata/1,
+  get_description/1,
+  get_tags/1,
+  set_limits/2
+]).
+
+-record(vhost, {
+          %% name as a binary
+          virtual_host,
+          %% proplist of limits configured, if any
+          limits,
+          metadata}).
+
+-define(record_version, vhost_v2).
+
+-type(name() :: binary()).
+
+-type vhost() :: vhost_v1:vhost_v1() | vhost_v2().
+
+-type vhost_v2() :: #vhost{
+                          virtual_host :: rabbit_types:vhost(),
+                          limits :: list(),
+                          metadata :: map() | undefined
+                         }.
+
+-type vhost_pattern() :: vhost_v1:vhost_v1_pattern() |
+                         vhost_v2_pattern().
+-type vhost_v2_pattern() :: #vhost{
+                                  virtual_host :: rabbit_types:vhost() | '_',
+                                  limits :: '_',
+                                  metadata :: '_'
+                                 }.
+
+-export_type([name/0,
+              vhost/0,
+              vhost_v2/0,
+              vhost_pattern/0,
+              vhost_v2_pattern/0]).
+
+-spec new(name(), list()) -> vhost().
+new(Name, Limits) ->
+  case record_version_to_use() of
+      ?record_version ->
+        #vhost{virtual_host = Name, limits = Limits};
+      _ ->
+        vhost_v1:new(Name, Limits)
+  end.
+
+-spec new(rabbit_vhost:name(), list(), map()) -> vhost().
+new(Name, Limits, Metadata) ->
+  case record_version_to_use() of
+      ?record_version ->
+        #vhost{virtual_host = Name, limits = Limits, metadata = Metadata};
+      _ ->
+        vhost_v1:new(Name, Limits)
+  end.
+
+-spec record_version_to_use() -> vhost_v1 | vhost_v2.
+
+record_version_to_use() ->
+    case rabbit_feature_flags:is_enabled(virtual_host_metadata) of
+        true  -> ?record_version;
+        false -> vhost_v1:record_version_to_use()
+    end.
+
+-spec upgrade(vhost()) -> vhost().
+
+upgrade(#vhost{} = VHost) -> VHost;
+upgrade(OldVHost)         -> upgrade_to(record_version_to_use(), OldVHost).
+
+-spec upgrade_to
+(vhost_v2, vhost()) -> vhost_v2();
+(vhost_v1, vhost_v1:vhost_v1()) -> vhost_v1:vhost_v1().
+
+upgrade_to(?record_version, #vhost{} = VHost) ->
+    VHost;
+upgrade_to(?record_version, OldVHost) ->
+    Fields = erlang:tuple_to_list(OldVHost) ++ [#{}],
+    #vhost{} = erlang:list_to_tuple(Fields);
+upgrade_to(Version, OldVHost) ->
+    vhost_v1:upgrade_to(Version, OldVHost).
+
+
+fields() ->
+    case record_version_to_use() of
+        ?record_version -> fields(?record_version);
+        _               -> vhost_v1:fields()
+    end.
+
+fields(?record_version) -> record_info(fields, vhost);
+fields(Version)         -> vhost_v1:fields(Version).
+
+info_keys() ->
+    case record_version_to_use() of
+        %% note: this reports description and tags separately even though
+        %% they are stored in the metadata map. MK.
+        ?record_version -> [name, description, tags, metadata, tracing, cluster_state];
```

Comment:
Should we instead use `metadata` as the info key and let the caller take whatever it wants from the returned map?

Reply:
We do report the metadata key but we also report `description` and `tags` so that formatting is easy on the CLI side.